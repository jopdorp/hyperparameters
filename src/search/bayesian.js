import * as math from 'mathjs';
import cho from 'cholesky';
import BaseSpace from '../base/base';
import RandomState from '../utils/RandomState';

// Implement the surrogate model using Gaussian Process Regression
class SurrogateModel {
  constructor() {
    this.X = [];
    this.y = [];
    this.noise = 1e-6;
    this.lengthScale = 1.0;
  }

  kernel(X1, X2) {
    // Ensure inputs are 2D arrays
    const x1Matrix = Array.isArray(X1[0]) ? X1 : X1.map(x => (Array.isArray(x) ? x : [x]));
    const x2Matrix = Array.isArray(X2[0]) ? X2 : X2.map(x => (Array.isArray(x) ? x : [x]));

    return x1Matrix.map(x1 =>
      x2Matrix.map((x2) => {
        // Ensure arrays have same dimensions before subtracting
        const x1Arr = Array.isArray(x1) ? x1 : [x1];
        const x2Arr = Array.isArray(x2) ? x2 : [x2];

        if (x1Arr.length !== x2Arr.length) {
          const maxLen = Math.max(x1Arr.length, x2Arr.length);
          while (x1Arr.length < maxLen) x1Arr.push(0);
          while (x2Arr.length < maxLen) x2Arr.push(0);
        }

        const diff = math.subtract(math.matrix(x1Arr), math.matrix(x2Arr));
        const squaredDist = math.sum(math.square(diff));
        return Math.exp((-0.5 * squaredDist) / (this.lengthScale ** 2));
      }));
  }

  fit(X, y) {
    if (!X.length || !y.length) return;

    this.X = X;
    this.y = Array.isArray(y) ? y : [y];

    const K = this.kernel(X, X);
    const kY = math.add(K, math.multiply(this.noise, math.identity(K.length)));

    try {
      // Use the standalone cholesky package
      this.L = cho(kY);
      // Convert L to the format expected by subsequent operations
      this.L = this.L.map((row) => {
        const fullRow = new Array(kY.length).fill(0);
        row.forEach((val, j) => {
          fullRow[j] = val;
        });
        return fullRow;
      });
      const yMatrix = math.matrix(this.y);
      this.alpha = math.lusolve(math.matrix(this.L), math.transpose(yMatrix));
    } catch (e) {
      // Fallback to identity matrix if decomposition fails
      this.L = math.identity(K.length);
      this.alpha = math.matrix(this.y);
    }
  }

  predict(XNew) {
    const kS = this.kernel(XNew, this.X);
    const kSS = this.kernel(XNew, XNew);
    const mu = kS.map(kSRow => math.multiply(kSRow, this.alpha));
    const v = math.lusolve(this.L, math.transpose(kS));
    const cov = math.subtract(kSS, math.multiply(math.transpose(v), v));
    const stdv = cov.map((row, i) => Math.sqrt(Math.max(row[i], 1e-9)));
    return { mean: mu.map(m => m[0]), stdv };
  }
}

// Implement the Expected Improvement acquisition function
function expectedImprovement(mu, sigma, yBest) {
  const epsilon = 1e-9;
  return mu.map((m, i) => {
    const s = sigma[i] + epsilon;
    const z = (yBest - m) / s;
    const phi = math.exp(-0.5 * (z * z)) / math.sqrt(2 * Math.PI);
    const Phi = 0.5 * (1.0 + math.erf(z / math.sqrt(2)));
    return s * ((z * Phi) + phi);
  });
}

class BayesianSearch extends BaseSpace {
  constructor() {
    super();
    this.surrogateModel = new SurrogateModel();
    this.observations = [];
    this.categoricalMaps = new Map(); // Store mappings for categorical variables
  }

  // Add all parameter handlers to match RandomSearch
  choice = (params, rng) => {
    const { options } = params;
    const idx = rng.randrange(0, options.length, 1);
    const option = options[idx];
    return this.eval(option, { rng });
  };

  randint = (params, rng) => rng.randrange(0, params.upper, 1);

  loguniform = (params, rng) => {
    const { low, high } = params;
    return Math.exp(rng.uniform(low, high));
  };

  qloguniform = (params, rng) => {
    const { low, high, q } = params;
    return Math.round(Math.exp(rng.uniform(low, high)) / q) * q;
  };

  normal = (params, rng) => {
    const { mu, sigma } = params;
    return rng.gauss(mu, sigma);
  };

  qnormal = (params, rng) => {
    const { mu, sigma, q } = params;
    return Math.round(rng.gauss(mu, sigma) / q) * q;
  };

  lognormal = (params, rng) => {
    const { mu, sigma } = params;
    return Math.exp(rng.gauss(mu, sigma));
  };

  qlognormal = (params, rng) => {
    const { mu, sigma, q } = params;
    return Math.round(Math.exp(rng.gauss(mu, sigma)) / q) * q;
  };

  // Update the surrogate model with new observations
  updateSurrogateModel() {
    const X = this.observations.map(obs => this.flattenParams(obs.params));
    const y = this.observations.map(obs => obs.result.loss || obs.result.accuracy);
    if (X.length > 0) {
      this.surrogateModel.fit(X, y);
    }
  }

  // Generate candidate points from the search space
  generateCandidates(expr, rng, numCandidates = 10) {
    const candidates = [];
    const flatCandidates = [];
    for (let i = 0; i < numCandidates; i += 1) {
      const candidate = super.eval(expr, { rng });
      candidates.push(candidate);
      flatCandidates.push(this.flattenParams(candidate));
    }
    return { candidates, flatCandidates };
  }

  // Helper method to encode categorical values
  encodeCategorical(value) {
    if (typeof value === 'string') {
      if (!this.categoricalMaps.has(value)) {
        this.categoricalMaps.set(value, this.categoricalMaps.size);
      }
      return this.categoricalMaps.get(value);
    }
    if (typeof value === 'boolean') {
      return value ? 1 : 0;
    }
    return value;
  }

  // Helper method to flatten parameter objects
  flattenParams(params) {
    if (Array.isArray(params)) {
      return params.flatMap(p => this.flattenParams(p));
    }
    if (typeof params === 'object' && params !== null) {
      return Object.values(params).flatMap(p => this.flattenParams(p));
    }
    return [this.encodeCategorical(params)];
  }

  // Propose the next set of parameters to evaluate
  propose(expr, rng) {
    const { candidates, flatCandidates } = this.generateCandidates(expr, rng, 100);
    const { mean, stdv } = this.surrogateModel.predict(flatCandidates);

    const yBest = Math.min(...this.observations.map(obs => obs.result.loss || Infinity));
    const eiValues = expectedImprovement(mean, stdv, yBest);

    const maxIndex = eiValues.indexOf(Math.max(...eiValues));
    return candidates[maxIndex]; // Return the original structured candidate
  }

  // Evaluate the expression using Bayesian optimization
  eval(expr, { rng }) {
    if (this.observations.length > 0) {
      return this.propose(expr, rng);
    }
    // Random sample for the initial observation
    return super.eval(expr, { rng });
  }

  // Add parameter handlers similar to RandomSearch
  uniform = (params, rng) => {
    const { low, high } = params;
    return rng.uniform(low, high);
  };

  quniform = (params, rng) => {
    const { low, high, q } = params;
    return Math.round(rng.uniform(low, high) / q) * q;
  };

  // Add other parameter handlers (normal, lognormal, etc.) as needed
}

export const bayesianSample = (space, params = {}) => {
  const bs = new BayesianSearch();
  const args = bs.eval(space, params);
  if (Object.keys(args).length === 1) {
    const results = Object.keys(args).map(key => args[key]);
    return results.length === 1 ? results[0] : results;
  }
  return args;
};

export const bayesianSearch = (newIds, domain, trials, seed) => {
  const rng = new RandomState(seed);
  let rval = [];
  const bs = new BayesianSearch();

  // Update observations from past trials
  bs.observations = trials.trials.map(trial => ({
    params: trial.args,
    result: trial.result,
  }));
  if (bs.observations.length > 0) {
    bs.updateSurrogateModel();
  }

  newIds.forEach((newId) => {
    const paramsEval = bs.eval(domain.expr, { rng });
    const result = domain.newResult();
    rval = [...rval, ...trials.newTrialDocs([newId], [result], [paramsEval])];
  });
  return rval;
};
