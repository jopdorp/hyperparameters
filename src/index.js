import fmin from './base/fmin';
import { randomSample, randomSearch } from './search/random';
import { gridSample, gridSearch } from './search/grid';
import RandomState from './utils/RandomState';
import { bayesianSample, bayesianSearch } from './search/bayesian';

export * from './base/base';

export { fmin, RandomState };

export const choice = options => ({ name: 'choice', options });
export const randint = upper => ({ name: 'randint', upper });
export const uniform = (low, high) => ({ name: 'uniform', low, high });
export const quniform = (low, high, q) => ({
  name: 'quniform', low, high, q
});
export const loguniform = (low, high) => ({ name: 'loguniform', low, high });
export const qloguniform = (low, high, q) => ({
  name: 'qloguniform', low, high, q
});
export const normal = (mu, sigma) => ({ name: 'normal', mu, sigma });
export const qnormal = (mu, sigma, q) => ({
  name: 'qnormal', mu, sigma, q
});
export const lognormal = (mu, sigma) => ({ name: 'lognormal', mu, sigma });
export const qlognormal = (mu, sigma, q) => ({
  name: 'qlognormal', mu, sigma, q
});

export const search = {
  randomSearch,
  gridSearch,
  bayesianSearch,
};

export const sample = {
  randomSample,
  gridSample,
  bayesianSample
};
