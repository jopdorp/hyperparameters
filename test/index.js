import chai, { assert } from 'chai';
import snapshots from 'chai-snapshot-tests';
import * as hpjs from '../src';
import { GridSearch } from '../src/search/grid';

chai.use(snapshots(__filename));

const seededSample = (space) => hpjs.sample.randomSample(space, { rng: new hpjs.RandomState(12345) });

const floatSeededSample = (space) => hpjs.sample.randomSample(space, { rng: new hpjs.RandomState(12345) }).toFixed(7);

const objectToFixed = (obj) => Object.keys(obj).reduce((final, key) => {
  const value = typeof obj[key] === 'number' ? obj[key].toFixed(7) : obj[key];
  return { ...final, [key]: value};
}, {});

const randFMinSeeded = async (opt, space) => {
  const trials = await hpjs.fmin(opt, space, hpjs.search.randomSearch, 100, { rng: new hpjs.RandomState(12345) });
  return objectToFixed(trials.argmin);
}


describe('hpjs.choice.', () => {
  it('is a string', () => {
    const val = seededSample(hpjs.choice(['cat', 'dog']));
    assert.typeOf(val, 'string');
  });
  it('is one of the elements', () => {
    const val = seededSample(hpjs.choice(['cat', 'dog']));
    assert(['cat', 'dog'].indexOf(val) >= 0, 'val was actually ' + val);
  });
  it('picks a number', () => {
    const val = seededSample(hpjs.choice([1, 2, 3, 4]));
    assert(val === 4, 'val was actually: ' + val);
  });
});

describe('hpjs.randint.', () => {
  it('in range [0,5)', () => {
    const val  = seededSample(hpjs.randint(5));
    assert(val >= 0 && val < 5, `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('upper: negative', floatSeededSample(hpjs.randint(-2)));
    assert.snapshot('upper: 0', floatSeededSample(hpjs.randint(0)));
    assert.snapshot('upper: 1', floatSeededSample(hpjs.randint(1)));
    assert.snapshot('upper: 1000000', floatSeededSample(hpjs.randint(1000000)));
  });

});

describe('hpjs.uniform.', () => {
  it('between 0 and 1', () => {
    const val  = seededSample(hpjs.uniform(0, 1));
    assert(val >= 0 && val <= 1, `actual value ${val}`);
  });
  it('between -1 and 1', () => {
    const val  = seededSample(hpjs.uniform(-1, 1));
    assert(val >= -1 && val <= 1, `actual value ${val}`);
  });

  it('Snapshot tests', () => {
    assert.snapshot('uniform -1, 1', floatSeededSample(hpjs.uniform(-1, 1)));
    assert.snapshot('uniform -100000, -1', floatSeededSample(hpjs.uniform(-100000, -1)));
    assert.snapshot('uniform -1, -10', floatSeededSample(hpjs.uniform(-1, -10)));
    assert.snapshot('uniform 5, 1', floatSeededSample(hpjs.uniform(5, 1)));
    assert.snapshot('uniform 1, 1000000', floatSeededSample(hpjs.uniform(1, 1000000)));
    assert.snapshot('uniform 1, 1', floatSeededSample(hpjs.uniform(1, 1)));
  });
});

describe('hpjs.quniform.', () => {
  it('between 0 and 1', () => {
    const val  = seededSample(hpjs.quniform(0, 1, 0.2));
    assert(val >= 0 && val <= 1, `actual value ${val}`);
  });
  it('between -1 and 1, step 1', () => {
    const val  = seededSample(hpjs.quniform(-1, 1, 1));
    assert(val === -1 || val === 0 || val === 1, `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('quniform -1, 1, 0.1', floatSeededSample(hpjs.quniform(-1, 1, 0.1)));
    assert.snapshot('quniform -100000, -1, -1', floatSeededSample(hpjs.quniform(-100000, -1, -1)));
    assert.snapshot('quniform -1, -10, 0.22222', floatSeededSample(hpjs.quniform(-1, -10, 0.22222)));
    assert.snapshot('quniform 5, 1, -0.111', floatSeededSample(hpjs.quniform(5, 1, -0.111)));
    assert.snapshot('quniform 1, 1000000, 50', floatSeededSample(hpjs.quniform(1, 1000000, 50)));
    assert.snapshot('quniform 1, 1, 0.001', floatSeededSample(hpjs.quniform(1, 1, 0.001)));
  });
});

describe('hpjs.loguniform.', () => {
  it('between e^0 and e^1', () => {
    const low = 0;
    const high = 1;
    const val  = seededSample(hpjs.loguniform(low, high));
    assert(val >= Math.exp(low) && val <= Math.exp(high), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('loguniform -1, 1', floatSeededSample(hpjs.loguniform(-1, 1)));
    assert.snapshot('loguniform -100000, -1', floatSeededSample(hpjs.loguniform(-100000, -1)));
    assert.snapshot('loguniform -1, -10', floatSeededSample(hpjs.loguniform(-1, -10)));
    assert.snapshot('loguniform 5, 1', floatSeededSample(hpjs.loguniform(5, 1)));
    assert.snapshot('loguniform 5, 1', floatSeededSample(hpjs.loguniform(5, 1)));
    assert.snapshot('loguniform 1, 1', floatSeededSample(hpjs.loguniform(1, 1)));
  });
});

describe('hpjs.qloguniform.', () => {
  it('e^0 and e^1', () => {
    const low = 0;
    const high = 1;
    const val  = seededSample(hpjs.qloguniform(low, high, 0.2));
    assert(val >= Math.exp(low) && val <= Math.exp(high), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('qloguniform -1, 1, 0.1', floatSeededSample(hpjs.qloguniform(-1, 1, 0.1)));
    assert.snapshot('qloguniform -100000, -1, -1', floatSeededSample(hpjs.qloguniform(-100000, -1, -1)));
    assert.snapshot('qloguniform -1, -10, 0.2222', floatSeededSample(hpjs.qloguniform(-1, -10, 0.22222)));
    assert.snapshot('qloguniform 5, 1, -0.111', floatSeededSample(hpjs.qloguniform(5, 1, -0.111)));
    assert.snapshot('qloguniform 5, 1, 0.1', floatSeededSample(hpjs.qloguniform(5, 1, 0.1)));
    assert.snapshot('qloguniform 1, 1, 0.001', floatSeededSample(hpjs.qloguniform(1, 1, 0.001)));
  });
});

describe('hpjs.normal.', () => {
  it('a number', () => {
    const mu = -1;
    const sigma = 1;
    const val  = seededSample(hpjs.normal(mu, sigma));
    assert(!isNaN(val), `actual value ${val}`);
  });
  it('within 3 standard deviations of mean', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hpjs.normal(mu, sigma));
    assert(val >= mu - (3*sigma) && val <= mu + (3*sigma), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('normal -1, 1', floatSeededSample(hpjs.normal(-1, 1)));
    assert.snapshot('normal -100000, -1', floatSeededSample(hpjs.normal(-100000, -1)));
    assert.snapshot('normal -1, -10', floatSeededSample(hpjs.normal(-1, -10)));
    assert.snapshot('normal 5, 1', floatSeededSample(hpjs.normal(5, 1)));
    assert.snapshot('normal 5, 1', floatSeededSample(hpjs.normal(5, 1)));
    assert.snapshot('normal 1, 1', floatSeededSample(hpjs.normal(1, 1)));
  });
});

describe('hpjs.qnormal.', () => {
  it('a number', () => {
    const mu = -1;
    const sigma = 1;
    const val  = seededSample(hpjs.normal(mu, sigma));
    assert(!isNaN(val), `actual value ${val}`);
  });
  it('within 3 standard deviations of mean', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hpjs.qnormal(mu, sigma, 0.1));
    assert(val >= mu - (3*sigma) && val <= mu + (3*sigma), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('qnormal -1, 1, 0.1', floatSeededSample(hpjs.qnormal(-1, 1, 0.1)));
    assert.snapshot('qnormal -100000, -1, -1', floatSeededSample(hpjs.qnormal(-100000, -1, -1)));
    assert.snapshot('qnormal -1, -10, 0.22222', floatSeededSample(hpjs.qnormal(-1, -10, 0.22222)));
    assert.snapshot('qnormal 5, 1, -0.111', floatSeededSample(hpjs.qnormal(5, 1, -0.111)));
    assert.snapshot('qnormal 1, 1000000, 50', floatSeededSample(hpjs.qnormal(1, 100, 50)));
    assert.snapshot('qnormal 1, 1, 0.001', floatSeededSample(hpjs.qnormal(1, 1, 0.001)));
  });
});

describe('hpjs.lognormal.', () => {
  it('positive', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hpjs.lognormal(mu, sigma));
    assert(val >= 0, `actual value ${val}`);
  });
  it('less ~e^3 from the mean, or less than ~3 standard deviations from it', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hpjs.lognormal(mu, sigma));
    assert(val <= 50, `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('lognormal -1, 1', floatSeededSample(hpjs.lognormal(-1, 1)));
    assert.snapshot('lognormal -100000, -1', floatSeededSample(hpjs.lognormal(-100000, -1)));
    assert.snapshot('lognormal -1, -10', floatSeededSample(hpjs.lognormal(-1, -10)));
    assert.snapshot('lognormal 5, 1', floatSeededSample(hpjs.lognormal(5, 1)));
    assert.snapshot('lognormal 5, 1', floatSeededSample(hpjs.lognormal(5, 1)));
    assert.snapshot('lognormal 1, 1', floatSeededSample(hpjs.lognormal(1, 1)));
  });
});

describe('hpjs.qlognormal.', () => {
  it('a number', () => {
    const mu = -1;
    const sigma = 1;
    const val  = seededSample(hpjs.qlognormal(mu, sigma, 0.2));
    assert(!isNaN(val), `actual value ${val}`);
  });
  it('within 3 standard deviations of mean', () => {
    const mu = 0;
    const sigma = 1;
    const val  = seededSample(hpjs.qlognormal(mu, sigma, 0.1));
    assert(val >= mu - (3*sigma) && val <= mu + (3*sigma), `actual value ${val}`);
  });
  it('Snapshot tests', () => {
    assert.snapshot('qlognormal -1, 1, 0.1', floatSeededSample(hpjs.qlognormal(-1, 1, 0.1)));
    assert.snapshot('qlognormal -100000, -1, -1', floatSeededSample(hpjs.qlognormal(-100000, -1, -1)));
    assert.snapshot('qlognormal -1, -10, 0.22222', floatSeededSample(hpjs.qlognormal(-1, -10, 0.22222)));
    assert.snapshot('qlognormal 5, 1, -0.111', floatSeededSample(hpjs.qlognormal(5, 1, -0.111)));
    assert.snapshot('qlognormal 5, 1, 0.1', floatSeededSample(hpjs.qlognormal(5, 1, 0.1)));
    assert.snapshot('qlognormal 1, 1, 0.001', floatSeededSample(hpjs.qlognormal(1, 1, 0.001)));
  });
});

describe('random sample', () => {
  it('Choice as array', () => {
    const space = hpjs.choice(
      [
        hpjs.lognormal(0, 1),
        hpjs.uniform(-10, 10)
      ]
    );

    assert.snapshot('sample: array', floatSeededSample(space));
  });
  it('more complex space with depth', () => {
    const space = {
      x: hpjs.normal(0, 2),
      y: hpjs.uniform(0, 1),
      choice: hpjs.choice([
        null, hpjs.uniform(0, 1),
      ]),
      array: [
        hpjs.normal(0, 2), hpjs.uniform(0, 3), hpjs.choice([false, true]),
      ],
      obj: {
        u: hpjs.uniform(0, 3),
        v: hpjs.uniform(0, 3),
        w: hpjs.uniform(-3, 0)
      }
    };
    assert.snapshot('sample: depth', objectToFixed(seededSample(space)));
  });
});
describe('grid search', () => {
   const gs = new GridSearch();
  it('choice', () => {
    const space = hpjs.choice(['cat', 'dog']);
    assert(gs.numSamples(space) === 2, `choice num samples ${gs.numSamples(space)}`);
  });
  it('randint', () => {
    const space = hpjs.randint(5);
    assert(gs.numSamples(space) === 5, `randint [0, 5) ${gs.numSamples(space)}`);
  });
  it('quniform', () => {
    let space = hpjs.quniform(0, 1, 0.1);
    assert(gs.numSamples(space) === 11, `quniform 0,1,0.1 ${gs.numSamples(space)}`);
    space = hpjs.quniform(-5, 5, 1);
    assert(gs.numSamples(space) === 11, `quniform -5,5,1 ${gs.numSamples(space)}`);
    space = hpjs.quniform(1, 10, 2);
    assert(gs.numSamples(space) === 5, `quniform 1,10,2 ${gs.numSamples(space)}`);
  });
  it('qloguniform', () => {
    let space = hpjs.qloguniform(0,1,0.1);
    assert(gs.numSamples(space) === 11, `qloguniform 0,1,0.1 ${gs.numSamples(space)}`);
    space = hpjs.qloguniform(-5,5,1);
    assert(gs.numSamples(space) === 11, `qloguniform -5,5,1 ${gs.numSamples(space)}`);
    space = hpjs.qloguniform(1,10,2);
    assert(gs.numSamples(space) === 5, `qloguniform 1,10,2 ${gs.numSamples(space)}`);
  });

  it('qnormal', () => {
    let space = hpjs.qnormal(0, 1, 0.1);
    assert(gs.numSamples(space) === 41, `qnormal 0,1,0.1 ${gs.numSamples(space)}`);
    space = hpjs.qnormal(-5,5,1);
    assert(gs.numSamples(space) === 21, `qnormal -5,5,1 ${gs.numSamples(space)}`);
    space = hpjs.qnormal(1,10,2);
    assert(gs.numSamples(space) === 21, `qnormal 1,10,2 ${gs.numSamples(space)}`);

  });
  it('uniform', () => {
    try {
      gs.numSamples(hpjs.uniform(0, 1));
      assert(false, 'hpjs.uniform not allowed for grid search');
    } catch (e) {
      assert(e.message === 'Can not evaluate length of non-discrete parameter "uniform"', `exception message ${e.message}`);
    }
  });
  it('loguniform', () => {
    try {
      gs.numSamples(hpjs.loguniform(0, 1));
      assert(false, 'hpjs.loguniform not allowed for grid search');
    } catch (e) {
      assert(e.message === 'Can not evaluate length of non-discrete parameter "loguniform"', `exception message ${e.message}`);
    }
  });
  it('normal', () => {
    try {
      gs.numSamples(hpjs.normal(0, 1));
      assert(false, 'hpjs.normal not allowed for grid search');
    } catch (e) {
      assert(e.message === 'Can not evaluate length of non-discrete parameter "normal"', `exception message ${e.message}`);
    }
  });
  it('lognormal', () => {
    try {
      gs.numSamples(hpjs.lognormal(0, 1));
      assert(false, 'hpjs.lognormal not allowed for grid search');
    } catch (e) {
      assert(e.message === 'Can not evaluate length of non-discrete parameter "lognormal"', `exception message ${e.message}`);
    }
  });
  it('choice grid search', () => {
    const space = hpjs.choice(
      [
        hpjs.qlognormal(0, 1, 1), //5
        hpjs.quniform(-10, 10, 1) //21
      ]
    );
    assert(gs.numSamples(space) === 26, `choice  ${gs.numSamples(space)}`);
  });
});

describe('fmin + rand', () => {
  it('FMin for x^2 - x + 1', async () => {
    const space = hpjs.uniform(-5, 5);
    const opt = x => ((x ** 2) - (x + 1));
    assert.snapshot('FMin for x^2 - x + 1', await randFMinSeeded(opt, space));
  });
  it('Hyperparameters space', async () => {
    const space = {
      x: hpjs.uniform(-5, 5),
      y: hpjs.uniform(-5, 5)
    };
    const opt = ({ x, y }) => ((x ** 2) + (y ** 2));
    assert.snapshot('Hyperparameters space', await randFMinSeeded(opt, space));
  });
  it('Choice selection of expressions', async () => {
    const space = hpjs.choice([
        hpjs.lognormal(0, 1),
        hpjs.uniform(-10, 10)
      ]
    );
    const opt = ( x ) => (x ** 2);
    assert.snapshot('choice as array space', await randFMinSeeded(opt, space));
  });
  it('Deep learning space', async () => {
    const space = {
      // Learning rate should be between 0.00001 and 1
      learning_rate:
        hpjs.loguniform(Math.log(1e-5), Math.log(1)),
      use_double_q_learning:
        hpjs.choice([true, false]),
      layer1_size: hpjs.quniform(10, 100, 1),
      layer2_size: hpjs.quniform(10, 100, 1),
      layer3_size: hpjs.quniform(10, 100, 1),
      future_discount_max: hpjs.uniform(0.5, 0.99),
      future_discount_increment: hpjs.loguniform(Math.log(0.001), Math.log(0.1)),
      recall_memory_size: hpjs.quniform(1, 100, 1),
      recall_memory_num_experiences_per_recall: hpjs.quniform(10, 2000, 1),
      num_epochs: hpjs.quniform(1, 10, 1),
    };

    const opt = params => params.learning_rate ** 2;

    assert.snapshot('Deep learning space', await randFMinSeeded(opt, space));
  });
});

describe('fmin + bayesian', () => {
  const bayesianFMinSeeded = async (opt, space) => {
    const trials = await hpjs.fmin(
      opt,
      space,
      hpjs.search.bayesianSearch,
      25,
      { rng: new hpjs.RandomState(12345) }
    );
    return objectToFixed(trials.argmin);
  };

  it('Simple quadratic function optimization', async () => {
    const space = hpjs.uniform(-5, 5);
    const opt = x => (x ** 2);
    assert.snapshot('Bayesian quadratic optimization', await bayesianFMinSeeded(opt, space));
  });

  it('2D quadratic optimization', async () => {
    const space = {
      x: hpjs.uniform(-5, 5),
      y: hpjs.uniform(-5, 5)
    };
    const opt = ({ x, y }) => (x ** 2) + (y ** 2);
    assert.snapshot('Bayesian 2D optimization', await bayesianFMinSeeded(opt, space));
  });

  it('Bayesian vs Random performance', async function() {
    this.timeout(10000);
    const space = {
      x: hpjs.uniform(-5, 5),
      y: hpjs.uniform(-5, 5)
    };
    const opt = ({ x, y }) => x ** 2 + y ** 2;
    
    const bayesianTrials = await hpjs.fmin(
      opt,
      space,
      hpjs.search.bayesianSearch,
      25,
      { rng: new hpjs.RandomState(12345) }
    );

    const randomTrials = await hpjs.fmin(
      opt,
      space,
      hpjs.search.randomSearch,
      25,
      { rng: new hpjs.RandomState(12345) }
    );

    const bayesianLoss = opt(bayesianTrials.argmin);
    const randomLoss = opt(randomTrials.argmin);

    assert(
      bayesianLoss <= randomLoss,
      `Bayesian (${bayesianLoss}) should perform better than Random (${randomLoss})`
    );
  });

  it('Complex hyperparameter space', async () => {
    const space = {
      learning_rate: hpjs.loguniform(Math.log(1e-5), Math.log(1)),
      num_layers: hpjs.choice([1, 2, 3]),
      layer_sizes: hpjs.choice([
        [hpjs.quniform(10, 50, 1)],
        [hpjs.quniform(10, 50, 1), hpjs.quniform(10, 50, 1)],
        [hpjs.quniform(10, 50, 1), hpjs.quniform(10, 50, 1), hpjs.quniform(10, 50, 1)]
      ]),
      activation: hpjs.choice(['relu', 'tanh', 'sigmoid']),
      dropout_rate: hpjs.uniform(0, 0.5)
    };

    const opt = params => {
      // Mock neural network performance metric
      return params.learning_rate * 0.1 + 
             (params.num_layers - 2) ** 2 + 
             params.dropout_rate;
    };

    assert.snapshot('Bayesian complex space', await bayesianFMinSeeded(opt, space));
  });

  it('Bayesian sample function', () => {
    const space = {
      x: hpjs.uniform(-5, 5),
      y: hpjs.quniform(0, 10, 1)
    };
    const sample = hpjs.sample.bayesianSample(space, { rng: new hpjs.RandomState(12345) });
    assert.snapshot('Bayesian sample', objectToFixed(sample));
  });
});

