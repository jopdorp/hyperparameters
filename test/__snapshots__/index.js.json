{
  "upper: negative": "0.0000000",
  "upper: 0": "0.0000000",
  "upper: 1": "0.0000000",
  "upper: 1000000": "933394.0000000",
  "uniform -1, 1": "0.7803094",
  "uniform -100000, -1": "-10985.4183013",
  "uniform -1, -10": "-9.0113925",
  "uniform 5, 1": "1.4393811",
  "uniform 1, 1000000": "890154.8283796",
  "uniform 1, 1": "1.0000000",
  "quniform -1, 1, 0.1": "0.8000000",
  "quniform -100000, -1, -1": "-10985.0000000",
  "quniform -1, -10, 0.22222": "-9.1110200",
  "quniform 5, 1, -0.111": "1.4430000",
  "quniform 1, 1000000, 50": "890150.0000000",
  "quniform 1, 1, 0.001": "1.0000000",
  "loguniform -1, 1": "2.1821474",
  "loguniform -100000, -1": "0.0000000",
  "loguniform -1, -10": "0.0001220",
  "loguniform 5, 1": "4.2180845",
  "loguniform 1, 1": "2.7182818",
  "qloguniform -1, 1, 0.1": "2.2000000",
  "qloguniform -100000, -1, -1": "0.0000000",
  "qloguniform -1, -10, 0.2222": "0.0000000",
  "qloguniform 5, 1, -0.111": "4.2180000",
  "qloguniform 5, 1, 0.1": "4.2000000",
  "qloguniform 1, 1, 0.001": "2.7180000",
  "normal -1, 1": "-0.3037819",
  "normal -100000, -1": "-100000.6962181",
  "normal -1, -10": "-7.9621809",
  "normal 5, 1": "5.6962181",
  "normal 1, 1": "1.6962181",
  "qnormal -1, 1, 0.1": "-0.3000000",
  "qnormal -100000, -1, -1": "-100001.0000000",
  "qnormal -1, -10, 0.22222": "-7.9999200",
  "qnormal 5, 1, -0.111": "5.6610000",
  "qnormal 1, 1000000, 50": "50.0000000",
  "qnormal 1, 1, 0.001": "1.6960000",
  "lognormal -1, 1": "0.7380218",
  "lognormal -100000, -1": "0.0000000",
  "lognormal -1, -10": "0.0003484",
  "lognormal 5, 1": "297.7392463",
  "lognormal 1, 1": "5.4532845",
  "qlognormal -1, 1, 0.1": "0.7000000",
  "qlognormal -100000, -1, -1": "0.0000000",
  "qlognormal -1, -10, 0.22222": "0.0000000",
  "qlognormal 5, 1, -0.111": "297.7020000",
  "qlognormal 5, 1, 0.1": "297.7000000",
  "qlognormal 1, 1, 0.001": "5.4530000",
  "sample: array": "-3.6724889",
  "sample: depth": {
    "x": "1.3924362",
    "y": "0.1307073",
    "choice": null,
    "array": [
      -0.6553433999116457,
      0.11927848398147867,
      false
    ],
    "obj": {
      "u": 2.47930839269838,
      "v": 1.703175088041835,
      "w": -1.4037662402075997
    }
  },
  "FMin for x^2 - x + 1": {},
  "Hyperparameters space": {
    "x": "-0.7975415",
    "y": "-0.7414791"
  },
  "choice as array space": {},
  "Deep learning space": {
    "learning_rate": "0.0000123",
    "use_double_q_learning": true,
    "layer1_size": "29.0000000",
    "layer2_size": "41.0000000",
    "layer3_size": "56.0000000",
    "future_discount_max": "0.6085554",
    "future_discount_increment": "0.0013308",
    "recall_memory_size": "8.0000000",
    "recall_memory_num_experiences_per_recall": "1275.0000000",
    "num_epochs": "2.0000000"
  },
  "Bayesian sample": {
    "x": "3.9015472",
    "y": "3.0000000"
  },
  "Bayesian quadratic optimization": {},
  "Bayesian 2D optimization": {
    "x": "1.2665640",
    "y": "-0.8930757"
  },
  "Bayesian complex space": {
    "learning_rate": "0.0000123",
    "num_layers": "2.0000000",
    "layer_sizes": [
      24
    ],
    "activation": "sigmoid",
    "dropout_rate": "0.1107709"
  }
}